src/racer/nameres.rs

```diff
@@ -2488,8 +2500,70 @@ pub(crate) fn get_index_output(selfm: &Match, session: &Session) -> Option<Ty> {
         .next()
         .and_then(|(_, item_ty)| match item_ty {
             Ty::PathSearch(paths) => {
-                get_assoc_type_from_header(&paths.path, selfm, &index_header, session)
+                get_assoc_type_from_header(&paths.path, context, impl_header, session)
             }
             _ => Some(item_ty),
         })
 }
+
+/// Checks if trait_impl is the impl TraitName<OtherType>
+fn has_impl_for_other_type(
+    trait_impl: &ImplHeader,
+    trait_name: &str,
+    other_type: Option<&str>,
+) -> bool {
+    if let Some(ref path) = trait_impl.trait_path() {
+        if path.name() == Some(trait_name) {
+            if other_type.is_none() && path.segments[0].generics.len() == 0 {
+                return true;
+            }
+            if path.segments[0].generics.len() > 0 {
+                return match path.segments[0].generics[0] {
+                    Ty::PathSearch(ref generic) => other_type == generic.path.name(),
+                    _ => false,
+                };
+            }
+            // default is self
+            return trait_impl.self_path().name() == other_type;
+        }
+    }
+    false
+}
+
+/// Resolves the type of a binary expression
+/// # Arguments
+/// * base_type: the type on the left hand side
+/// * node: the operator
+/// * other_type: the type on the right hand side
+/// *Note*: Works only for std::ops::* now
+pub(crate) fn resolve_binary_expr_type(
+    base_type: &Match,
+    node: BinOpKind,
+    other_type: Option<&str>,
+    session: &Session,
+) -> Option<Ty> {
```

Comment:
Now this function doesn't fail so should return `Match`

Reply:
But let's keep it option. We don't need to refactor if we enhance in the future

Reply:
> We don't need to refactor if we enhance in the future

I don't feel it 'enhance' to return `Option` here

Reply:
What if you could always find an impl and there isn't a valid impl for this type? You have to return `None`. That said, I am not against returning just a `Match`

Reply:
> What if you could always find an impl

I think it's very hard... but I also feel that  we shouldn't always returns fallback(base_type) too.

Let's leave it as it is now, and let's change it if we come up with a better idea.